src/charm.py

```diff
@@ -103,7 +117,119 @@ def _on_stop(self, _: ops.StopEvent) -> None:
+        """Handle leader-elected event."""
+        # We check that we are still the leader when starting to process this event
+        if self.unit.is_leader():
+            self._become_active()
+
+    def _on_peer_relation_departed(self, _: ops.RelationDepartedEvent) -> None:
+        """Handle the peer relation departed event."""
+        # We check that we are still the leader when starting to process this event
+        if self.unit.is_leader():
+            self._become_active()
+
+    def _become_active(self) -> bool:
+        """Set the current unit as the active unit of the charm.
+
+        Returns:
+            True if the charm is effectively the new active unit.
+        """
+        active_unit_ip = self._active_unit_ip()
+
+        assert self.peer_relation is not None  # nosec
+        if not active_unit_ip:
+            self.peer_relation.data[self.app].update({"active-unit": self._unit_ip()})
+            return True
+
+        if active_unit_ip != self._unit_ip():
+            status = self._dig_query(
+                f"@{active_unit_ip} service.{constants.ZONE_SERVICE_NAME} TXT +short",
+                retry=True,
+                wait=1,
+            )
+            if status != "ok":
+                self.peer_relation.data[self.app].update({"active-unit": self._unit_ip()})
+                return True
+            return False
+
+        return True
+
+    async def _dig_query(self, cmd: str, retry: bool = False, wait: int = 5) -> str:
+        """Query a DnsEntry with dig.
+
+        This function was created for simplicity's sake. If we need to make more DNS requests
+        in the future, we should revisit it by employing a python library.
+
+        Args:
+            cmd: The dig command to perform
+            retry: If the dig request should be retried
+            wait: duration in seconds to wait between retries
+
+        Returns: the result of the DNS query
+        """
+        result: str = ""
+        retry = False
+        for _ in range(5):
+            result = str(
+                subprocess.run(
+                    ["dig"] + cmd.split(" "),  # nosec
+                    capture_output=True,
+                    text=True,
+                    check=True,
+                )
+            ).strip()
+            if result != "" or not retry:
+                break
+            time.sleep(wait)
+
+        return result
+
+    def _is_active_unit(self) -> bool:
+        """Check if the charm is the active unit.
+
+        Returns:
+            True if the charm is effectively the active unit.
+        """
+        return self._active_unit_ip() == self._unit_ip()
+
+    def _active_unit_ip(self) -> str:
+        """Get current active unit ip.
+
+        Returns:
+            The IP of the active unit
+
+        Raises:
+            PeerRelationUnavailableError: when the peer relation does not exist
+        """
+        if not self.peer_relation:
```

Comment:
Does this mean we won't be able to deploy the charm with a single unit?

Reply:
The peer relation exists even if there is only one unit